homeassistant/components/tradfri/diagnostics.py

```diff
@@ -0,0 +1,45 @@
+"""Diagnostics support for IKEA Tradfri."""
+from __future__ import annotations
+
+from pytradfri import Gateway, PytradfriError
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+
+from .const import (
+    CONF_GATEWAY_ID,
+    COORDINATOR,
+    COORDINATOR_LIST,
+    DOMAIN,
+    GROUPS_LIST,
+    KEY_API,
+    TIMEOUT_API,
+)
+
+
+async def async_get_config_entry_diagnostics(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> dict:
+    """Return diagnostics the Tradfri platform."""
+    _entry_data = hass.data[DOMAIN][entry.entry_id]
+
+    _coordinator_data = _entry_data[COORDINATOR]
+
+    _api = _coordinator_data[KEY_API]
+    _gateway: Gateway = _coordinator_data[CONF_GATEWAY_ID]
+
+    try:
+        _gateway_info = await _api(_gateway.get_gateway_info(), timeout=TIMEOUT_API)
+        _fw_version = _gateway_info.firmware_version
```

Comment:
Get the gateway device entry from the device registry via the gateway id.

https://github.com/home-assistant/core/blob/c93d389544e075fbb567acfe26067a96730963e2/homeassistant/helpers/device_registry.py#L234-L240

When we have the device entry, the software version is one of its attributes.